### PR TITLE
feat: add app.configureHostResolver

### PR DESCRIPTION
**Pull request by nornagon**
#### Description of Change
This adds a new API, `app.configureHostResolver`, and also configures the host
resolver slightly differently by default:

1. The built-in resolver is used in preference to `getaddrinfo` when the
   `AsyncDns` feature flag is enabled, which it is by default on macOS.
2. DNS-over-HTTPS is enabled by default, which will only have an effect when
   the system provider is one of the known-to-Chromium DoH providers.
   Additionally, this respects the `DnsOverHttps` feature flag, along with the
   `Fallback` param, e.g. `--enable-features=DnsOverHttps:Fallback/false` to
   disable fallback to insecure DNS.
3. DoH servers can be specified on the command line using the `Templates`
   feature param, e.g. `--enable-features=DnsOverHttps:Templates/https%3A%2F%2Fdoh.opendns.com%2Fdns-query`

The command-line functionality is left undocumented as (1) it matches Chrome
and (2) it's unclear how stable that feature is in Chrome.

Ref https://github.com/electron/electron/issues/21064.

#### Checklist

- [x] PR description included and stakeholders cc'd
- [x] `npm test` passes
- [x] tests are [changed or added](https://github.com/electron/electron/blob/master/docs/development/testing.md)
- [x] relevant documentation is changed or added
- [x] [PR release notes](https://github.com/electron/clerk/blob/master/README.md) describe the change in a way relevant to app developers, and are [capitalized, punctuated, and past tense](https://github.com/electron/clerk/blob/master/README.md#examples).

#### Release Notes

Notes: Added `app.configureHostResolver` API for configuring DNS-over-HTTPS.

### CODE REVIEW

File: docs/api/app.md [hunk 1]
```diff
@@ -1061,6 +1061,54 @@ Imports the certificate in pkcs12 format into the platform certificate store.
...
+    resolver is used in preference to getaddrinfo. When enabled, the built-in
+    resolver will attempt to use the system's DNS settings to do DNS lookups
+    itself. Enabled by default on macOS, disabled by default on Windows and
+    Linux.
+  * `secureDnsMode` String (optional) - Can be "off", "automatic" or "secure".
+    Configures the DNS-over-HTTP mode. When "off", no DoH lookups will be
+    performed. When "automatic", DoH lookups will be peformed first if DoH is
+    available, and insecure DNS lookups will be performed as a fallback. When
+    "secure", only DoH lookups will be performed. Defaults to "automatic".
+  * `secureDnsServers` String[]&#32;(optional) - A list of DNS-over-HTTP
```
Comment (jkleinsc) [inline docs/api/app.md:1077]:
```suggestion
  * `secureDnsServers` String[] (optional) - A list of DNS-over-HTTP
```

**Comment by jkleinsc** [comment]
API LGTM

**Comment by codebytere** [comment]
API LGTM

**Review by codebytere — APPROVED** [review]

**Comment by nornagon** [comment]
CI failures are unrelated. Merging.

File: docs/api/app.md [hunk 2]
```diff
@@ -1061,6 +1061,61 @@ Imports the certificate in pkcs12 format into the platform certificate store.
...
+
+1. DNS-over-HTTPS, if the [DNS provider supports it][doh-providers], then
+2. the built-in resolver (enabled on macOS only by default), then
+3. the system's resolver (e.g. `getaddrinfo`).
+
+This can be configured to either restrict usage of non-encrypted DNS
+(`secureDnsMode: "secure"`), or disable DNS-over-HTTPS (`secureDnsMode:
+"off"`). It is also possible to enable or disable the built-in resolver.
+
+To disable insecure DNS, you specify a `secureDnsMode` of `"secure"`. If you do
```
Comment (nornagon) [inline docs/api/app.md:1100]:
```suggestion
To disable insecure DNS, you can specify a `secureDnsMode` of `"secure"`. If you do
```

**Pull request merged**